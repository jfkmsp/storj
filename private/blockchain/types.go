// Copyright (C) 2022 Storj Labs, Inc.
// See LICENSE for copying information.

package blockchain

import (
	"encoding/hex"
	"encoding/json"

	"github.com/zeebo/errs"
)

// Error for the package (mainly parsing errors).
var Error = errs.Class("blockchain")

// Lengths of hashes and addresses in bytes.
const (
	// HashLength is the expected length of the hash.
	HashLength = 32
	// AddressLength is the expected length of the address.
	AddressLength = 20
)

// Hash represents the 32 byte Keccak256 hash of arbitrary data.
type Hash [HashLength]byte

var _ json.Marshaler = Hash{}

// MarshalJSON implements json marshalling interface.
func (h Hash) MarshalJSON() ([]byte, error) {
	return json.Marshal(h.Hex())
}

// Bytes gets the byte representation of the underlying hash.
func (h Hash) Bytes() []byte { return h[:] }

// Hex gets the hex string representation of the underlying hash.
func (h Hash) Hex() string {
	return hex.EncodeToString(h.Bytes())
}

// Address is wallet address.
type Address [AddressLength]byte

var _ json.Marshaler = Address{}

// MarshalJSON implements json marshalling interface.
func (a Address) MarshalJSON() ([]byte, error) {
	return json.Marshal(a.Hex())
}

// Bytes gets the byte representation of the underlying address.
func (a Address) Bytes() []byte { return a[:] }

// Hex gets string representation of the underlying address.
func (a Address) Hex() string {
	return hex.EncodeToString(a.Bytes())
}
