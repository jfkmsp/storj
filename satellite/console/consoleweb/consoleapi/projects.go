// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

package consoleapi

import (
	"encoding/base64"
	"encoding/json"
	"go.opentelemetry.io/otel"
	"net/http"
	"os"

	"runtime"

	"github.com/gorilla/mux"
	"github.com/zeebo/errs"
	"go.uber.org/zap"

	"storj.io/common/uuid"
	"storj.io/storj/satellite/console"
)

// Projects is an api controller that exposes projects related functionality.
type Projects struct {
	log     *zap.Logger
	service *console.Service
}

// NewProjects is a constructor for api analytics controller.
func NewProjects(log *zap.Logger, service *console.Service) *Projects {
	return &Projects{
		log:     log,
		service: service,
	}
}

// GetSalt returns the project's salt.
func (p *Projects) GetSalt(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	var err error
	pc, _, _, _ := runtime.Caller(0)
	ctx, span := otel.Tracer(os.Getenv("SERVICE_NAME")).Start(ctx, runtime.FuncForPC(pc).Name())
	defer span.End()

	w.Header().Set("Content-Type", "application/json")

	idParam, ok := mux.Vars(r)["id"]
	if !ok {
		p.serveJSONError(w, http.StatusBadRequest, errs.New("missing id route param"))
		return
	}

	id, err := uuid.FromString(idParam)
	if err != nil {
		p.serveJSONError(w, http.StatusBadRequest, err)
	}

	salt, err := p.service.GetSalt(ctx, id)
	if err != nil {
		p.serveJSONError(w, http.StatusUnauthorized, err)
		return
	}

	b64SaltString := base64.StdEncoding.EncodeToString(salt)

	err = json.NewEncoder(w).Encode(b64SaltString)
	if err != nil {
		p.serveJSONError(w, http.StatusInternalServerError, err)
	}
}

// serveJSONError writes JSON error to response output stream.
func (p *Projects) serveJSONError(w http.ResponseWriter, status int, err error) {
	ServeJSONError(p.log, w, status, err)
}
