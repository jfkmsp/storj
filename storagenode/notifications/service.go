// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

package notifications

import (
	"context"
	"go.opentelemetry.io/otel"
	"os"

	"runtime"

	"go.uber.org/zap"

	"storj.io/common/uuid"
)

// TimesNotified is a numeric value of amount of notifications being sent to user.
type TimesNotified int

const (
	// TimesNotifiedZero haven't being notified yet.
	TimesNotifiedZero TimesNotified = 0
	// TimesNotifiedFirst sent notification one time.
	TimesNotifiedFirst TimesNotified = 1
	// TimesNotifiedSecond sent notifications twice.
	TimesNotifiedSecond TimesNotified = 2
	// TimesNotifiedLast three notification has been send.
	TimesNotifiedLast TimesNotified = 3
)

// Service is the notification service between storage nodes and satellites.
// architecture: Service
type Service struct {
	log *zap.Logger
	db  DB
}

// NewService creates a new notification service.
func NewService(log *zap.Logger, db DB) *Service {
	return &Service{
		log: log,
		db:  db,
	}
}

// Receive - receives notifications from satellite and Insert them into DB.
func (service *Service) Receive(ctx context.Context, newNotification NewNotification) (Notification, error) {
	notification, err := service.db.Insert(ctx, newNotification)
	if err != nil {
		return Notification{}, err
	}

	return notification, nil
}

// Read - change notification status to Read by ID.
func (service *Service) Read(ctx context.Context, notificationID uuid.UUID) (err error) {
	pc, _, _, _ := runtime.Caller(0)
	ctx, span := otel.Tracer(os.Getenv("SERVICE_NAME")).Start(ctx, runtime.FuncForPC(pc).Name())
	defer span.End()

	err = service.db.Read(ctx, notificationID)
	if err != nil {
		return err
	}

	return nil
}

// ReadAll - change status of all user's notifications to Read.
func (service *Service) ReadAll(ctx context.Context) (err error) {
	pc, _, _, _ := runtime.Caller(0)
	ctx, span := otel.Tracer(os.Getenv("SERVICE_NAME")).Start(ctx, runtime.FuncForPC(pc).Name())
	defer span.End()

	err = service.db.ReadAll(ctx)
	if err != nil {
		return err
	}

	return nil
}

// List - shows the list of paginated notifications.
func (service *Service) List(ctx context.Context, cursor Cursor) (_ Page, err error) {
	pc, _, _, _ := runtime.Caller(0)
	ctx, span := otel.Tracer(os.Getenv("SERVICE_NAME")).Start(ctx, runtime.FuncForPC(pc).Name())
	defer span.End()

	notificationPage, err := service.db.List(ctx, cursor)
	if err != nil {
		return Page{}, err
	}

	if notificationPage.Notifications == nil {
		notificationPage = Page{Notifications: []Notification{}}
	}

	return notificationPage, nil
}

// UnreadAmount - returns amount on notifications with value is_read = nil.
func (service *Service) UnreadAmount(ctx context.Context) (_ int, err error) {
	pc, _, _, _ := runtime.Caller(0)
	ctx, span := otel.Tracer(os.Getenv("SERVICE_NAME")).Start(ctx, runtime.FuncForPC(pc).Name())
	defer span.End()

	amount, err := service.db.UnreadAmount(ctx)
	if err != nil {
		return 0, err
	}

	return amount, nil
}
