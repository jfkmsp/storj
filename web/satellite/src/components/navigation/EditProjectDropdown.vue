// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="edit-project">
        <div class="edit-project__selection-area" :class="{ active: isDropdownShown, 'on-edit': isEditPage }" @click.stop.prevent="toggleDropdown">
            <p class="edit-project__selection-area__name" :title="projectName">{{ projectName }}</p>
            <DotsImage class="edit-project__selection-area__image" />
        </div>
        <div v-if="isDropdownShown" v-click-outside="closeDropdown" class="edit-project__dropdown">
            <div class="edit-project__dropdown__choice" @click.stop.prevent="onEditProjectClick">
                <EditImage />
                <p class="edit-project__dropdown__choice__label">Edit Details</p>
            </div>
        </div>
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import DotsImage from '@/../static/images/navigation/dots.svg';
import EditImage from '@/../static/images/navigation/edit.svg';

import { RouteConfig } from '@/router';
import { APP_STATE_ACTIONS } from '@/utils/constants/actionNames';

// @vue/component
@Component({
    components: {
        DotsImage,
        EditImage,
    },
})
export default class EditProjectDropdown extends Vue {
    /**
     * Returns selected project's name.
     */
    public get projectName(): string {
        return this.$store.getters.selectedProject.name;
    }

    /**
     * Indicates if dropdown is shown.
     */
    public get isDropdownShown(): string {
        return this.$store.state.appStateModule.appState.isEditProjectDropdownShown;
    }

    /**
     * Indicates if current route name equals "edit project details" route name.
     */
    public get isEditPage(): boolean {
        return this.$route.name === RouteConfig.EditProjectDetails.name;
    }

    /**
     * Redirects to edit project details page.
     */
    public onEditProjectClick(): void {
        this.closeDropdown();
        this.$router.push(RouteConfig.EditProjectDetails.path);
    }

    /**
     * Toggles dropdown visibility.
     */
    public toggleDropdown(): void {
        this.$store.dispatch(APP_STATE_ACTIONS.TOGGLE_EDIT_PROJECT_DROPDOWN);
    }

    /**
     * Closes dropdown.
     */
    public closeDropdown(): void {
        if (!this.isDropdownShown) return;

        this.$store.dispatch(APP_STATE_ACTIONS.CLOSE_POPUPS);
    }
}
</script>

<style scoped lang="scss">
    .edit-project {
        font-family: 'font_regular', sans-serif;
        position: relative;
        width: 185px;
        margin: 0 0 30px 15px;

        &__selection-area {
            width: calc(100% - 15px);
            display: flex;
            align-items: center;
            justify-content: space-between;
            padding: 10px;
            border-radius: 6px;
            cursor: pointer;

            &__name {
                font-family: 'font_medium', sans-serif;
                font-size: 18px;
                line-height: 22px;
                color: #000;
                margin: 0 5px 0 0;
                white-space: nowrap;
                overflow: hidden;
                text-overflow: ellipsis;
            }

            &__image {
                min-width: 3px;
            }
        }

        &__dropdown {
            position: absolute;
            top: calc(100% + 5px);
            left: 0;
            background-color: #fff;
            box-shadow: 0 8px 34px rgb(161 173 185 / 41%);
            border-radius: 6px;
            padding: 5px 0;
            width: calc(100% + 5px);
            z-index: 1;

            &__choice {
                background-color: #fff;
                width: calc(100% - 32px);
                padding: 10px 16px;
                cursor: pointer;
                display: flex;
                align-items: center;

                &__label {
                    margin: 0 0 0 10px;
                    font-weight: 500;
                    font-size: 14px;
                    line-height: 19px;
                    color: #354049;
                }

                &:hover {
                    background-color: #f5f5f7;

                    .edit-project__dropdown__choice__label {
                        font-weight: unset;
                        font-family: 'font_bold', sans-serif;
                    }
                }
            }
        }
    }

    .active {
        background-color: #fff;
    }

    .on-edit {
        background-color: #0068dc;

        .edit-project__selection-area__name {
            color: #fff;
        }

        .edit-dots-svg-path {
            fill: #fff;
        }
    }
</style>
