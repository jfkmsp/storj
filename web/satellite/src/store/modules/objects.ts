// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

import {
    S3Client,
    Bucket,
    ListBucketsCommand,
    CreateBucketCommand,
    DeleteBucketCommand
} from '@aws-sdk/client-s3';

import { EdgeCredentials } from '@/types/accessGrants';
import { APP_STATE_ACTIONS } from '@/utils/constants/actionNames';
import { FilesState } from '@/store/modules/files';
import {StoreModule} from "@/types/store";
import {SignatureV4} from "@aws-sdk/signature-v4";

export const OBJECTS_ACTIONS = {
    CLEAR: 'clearObjects',
    SET_GATEWAY_CREDENTIALS: 'setGatewayCredentials',
    SET_API_KEY: 'setApiKey',
    SET_S3_CLIENT: 'setS3Client',
    SET_PASSPHRASE: 'setPassphrase',
    SET_FILE_COMPONENT_BUCKET_NAME: 'setFileComponentBucketName',
    FETCH_BUCKETS: 'fetchBuckets',
    CREATE_BUCKET: 'createBucket',
    CREATE_DEMO_BUCKET: 'createDemoBucket',
    DELETE_BUCKET: 'deleteBucket',
    CHECK_ONGOING_UPLOADS: 'checkOngoingUploads',
};

export const OBJECTS_MUTATIONS = {
    SET_GATEWAY_CREDENTIALS: 'setGatewayCredentials',
    SET_API_KEY: 'setApiKey',
    CLEAR: 'clearObjects',
    SET_S3_CLIENT: 'setS3Client',
    SET_BUCKETS: 'setBuckets',
    SET_FILE_COMPONENT_BUCKET_NAME: 'setFileComponentBucketName',
    SET_PASSPHRASE: 'setPassphrase',
    SET_LEAVE_ROUTE: 'setLeaveRoute',
};

export const DEMO_BUCKET_NAME = 'demo-bucket';

const {
    CLEAR,
    SET_API_KEY,
    SET_GATEWAY_CREDENTIALS,
    SET_S3_CLIENT,
    SET_BUCKETS,
    SET_PASSPHRASE,
    SET_FILE_COMPONENT_BUCKET_NAME,
    SET_LEAVE_ROUTE,
} = OBJECTS_MUTATIONS;

export class ObjectsState {
    public apiKey = '';
    public gatewayCredentials: EdgeCredentials = new EdgeCredentials();
    public s3Client: S3Client = new S3Client({
        forcePathStyle: true,
        signerConstructor: SignatureV4,
        // TODO: httpOptions: { timeout: 0 },
    });
    public bucketsList: Bucket[] = [];
    public passphrase = '';
    public fileComponentBucketName = '';
    public leaveRoute = '';
}

interface ObjectsContext {
    state: ObjectsState
    commit: (string, ...unknown) => void
    dispatch: (string, ...unknown) => Promise<any> // eslint-disable-line @typescript-eslint/no-explicit-any
    rootState: {
        files: FilesState
    }
}

/**
 * Creates objects module with all dependencies.
 */
export function makeObjectsModule(): StoreModule<ObjectsState, ObjectsContext> {
    return {
        state: new ObjectsState(),
        mutations: {
            [SET_API_KEY](state: ObjectsState, apiKey: string) {
                state.apiKey = apiKey;
            },
            [SET_GATEWAY_CREDENTIALS](state: ObjectsState, credentials: EdgeCredentials) {
                state.gatewayCredentials = credentials;
            },
            [SET_S3_CLIENT](state: ObjectsState) {
                state.s3Client = new S3Client({
                    credentials: {
                        accessKeyId: state.gatewayCredentials.accessKeyId,
                        secretAccessKey: state.gatewayCredentials.secretKey,
                    },
                    endpoint: state.gatewayCredentials.endpoint,
                    forcePathStyle: true,
                    signerConstructor: SignatureV4,
                    // TODO: httpOptions: { timeout: 0 },
                });
            },
            [SET_BUCKETS](state: ObjectsState, buckets: Bucket[]) {
                state.bucketsList = buckets;
            },
            [SET_PASSPHRASE](state: ObjectsState, passphrase: string) {
                state.passphrase = passphrase;
            },
            [SET_FILE_COMPONENT_BUCKET_NAME](state: ObjectsState, bucketName: string) {
                state.fileComponentBucketName = bucketName;
            },
            [SET_LEAVE_ROUTE](state: ObjectsState, leaveRoute: string) {
                state.leaveRoute = leaveRoute;
            },
            [CLEAR](state: ObjectsState) {
                state.apiKey = '';
                state.passphrase = '';
                state.gatewayCredentials = new EdgeCredentials();
                state.s3Client = new S3Client({
                    forcePathStyle: true,
                    signerConstructor: SignatureV4,
                    // TODO: httpOptions: { timeout: 0 },
                });
                state.bucketsList = [];
                state.fileComponentBucketName = '';
            },
        },
        actions: {
            setApiKey: function({commit}: ObjectsContext, apiKey: string): void {
                commit(SET_API_KEY, apiKey);
            },
            setGatewayCredentials: function({commit}: ObjectsContext, credentials: EdgeCredentials): void {
                commit(SET_GATEWAY_CREDENTIALS, credentials);
            },
            setS3Client: function({commit}: ObjectsContext): void {
                commit(SET_S3_CLIENT);
            },
            setPassphrase: function({commit}: ObjectsContext, passphrase: string): void {
                commit(SET_PASSPHRASE, passphrase);
            },
            setFileComponentBucketName: function({commit}: ObjectsContext, bucketName: string): void {
                commit(SET_FILE_COMPONENT_BUCKET_NAME, bucketName);
            },
            fetchBuckets: async function(ctx): Promise<void> {
                const result = await ctx.state.s3Client.send(new ListBucketsCommand({}));
                ctx.commit(SET_BUCKETS, result.Buckets);
            },
            createBucket: async function(ctx, name: string): Promise<void> {
                await ctx.state.s3Client.send(new CreateBucketCommand({
                    Bucket: name,
                }))
            },
            createDemoBucket: async function(ctx): Promise<void> {
                await ctx.state.s3Client.send(new CreateBucketCommand({
                    Bucket: DEMO_BUCKET_NAME,
                }))
            },
            deleteBucket: async function(ctx, name: string): Promise<void> {
                await ctx.state.s3Client.send(
                    new DeleteBucketCommand({
                        Bucket: name,
                    }));
            },
            clearObjects: function({commit}: ObjectsContext): void {
                commit(CLEAR);
            },
            checkOngoingUploads: function({commit, dispatch, rootState}: ObjectsContext, leaveRoute: string): boolean {
                if (!rootState.files.uploading.length) {
                    return false;
                }

                commit(SET_LEAVE_ROUTE, leaveRoute);
                dispatch(APP_STATE_ACTIONS.TOGGLE_UPLOAD_CANCEL_POPUP, null, {root: true});

                return true;
            },
        },
    };
}
