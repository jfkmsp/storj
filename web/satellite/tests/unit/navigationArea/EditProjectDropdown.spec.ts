// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import { VNode } from 'vue';
import { DirectiveBinding } from 'vue/types/options';
import Vuex from 'vuex';

import EditProjectDropdown from '@/components/navigation/EditProjectDropdown.vue';

import { router } from '@/router';
import { appStateModule } from '@/store/modules/appState';
import { makeProjectsModule, PROJECTS_MUTATIONS } from '@/store/modules/projects';
import { Project } from '@/types/projects';
import { createLocalVue, mount } from '@vue/test-utils';

import { ProjectsApiMock } from '../mock/api/projects';
import { makeNotificationsModule } from "@/store/modules/notifications";
import { NotificatorPlugin } from "@/utils/plugins/notificator";

const localVue = createLocalVue();
localVue.use(Vuex);

const projectsApi = new ProjectsApiMock();
const projectsModule = makeProjectsModule(projectsApi);
const notificationsModule = makeNotificationsModule();
const store = new Vuex.Store({ modules: { projectsModule, appStateModule, notificationsModule }});
const project = new Project('id', 'test', 'test', 'test', 'ownedId', false);

localVue.use(new NotificatorPlugin(store));

let clickOutsideEvent: EventListener;

localVue.directive('cli' +
    'ck-outside', {
    bind: function (el: HTMLElement, binding: DirectiveBinding, vnode: VNode) {
        clickOutsideEvent = function(event: Event): void {
            if (el === event.target) {
                return;
            }

            if (vnode.context && binding.expression) {
                vnode.context[binding.expression](event);
            }
        };

        document.body.addEventListener('click', clickOutsideEvent);
    },
    unbind: function(): void {
        document.body.removeEventListener('click', clickOutsideEvent);
    },
});


store.commit(PROJECTS_MUTATIONS.ADD, project);
store.commit(PROJECTS_MUTATIONS.SELECT_PROJECT, project.id);

describe('EditProjectDropdown', () => {
    it('renders correctly', (): void => {
        const wrapper = mount(EditProjectDropdown, {
            store,
            localVue,
            router,
        });

        expect(wrapper).toMatchSnapshot();
    });

    it('dropdown opens correctly', async (): Promise<void> => {
        const wrapper = mount(EditProjectDropdown, {
            store,
            localVue,
            router,
        });

        await wrapper.find('.edit-project__selection-area').trigger('click');

        expect(wrapper).toMatchSnapshot();
    });
});
