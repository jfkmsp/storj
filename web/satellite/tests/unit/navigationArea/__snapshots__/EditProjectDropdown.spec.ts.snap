// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditProjectDropdown dropdown opens correctly 1`] = `
<div class="edit-project">
  <div class="edit-project__selection-area active">
    <p title="test" class="edit-project__selection-area__name">test</p> <svg class="edit-project__selection-area__image"></svg>
  </div>
  <div class="edit-project__dropdown">
    <div class="edit-project__dropdown__choice"><svg></svg>
      <p class="edit-project__dropdown__choice__label">Edit Details</p>
    </div>
  </div>
</div>
`;

exports[`EditProjectDropdown renders correctly 1`] = `
<div class="edit-project">
  <div class="edit-project__selection-area">
    <p title="test" class="edit-project__selection-area__name">test</p> <svg class="edit-project__selection-area__image"></svg>
  </div>
  <!---->
</div>
`;
